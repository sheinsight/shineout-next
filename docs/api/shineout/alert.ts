const api = JSON.parse('[{"title":"Alert","properties":[{"name":"className","tag":{"cn":"自定义类名","en":"Custom class name","default":"","version":""},"required":false,"type":"string "},{"name":"style","tag":{"cn":"自定义样式","en":"Custom style","default":"","version":""},"required":false,"type":"CSSProperties "}],"cn":"","en":""}]');
export default api;
