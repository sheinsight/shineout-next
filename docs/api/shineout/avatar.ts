const api = JSON.parse('[{"title":"Avatar","properties":[{"name":"className","tag":{"cn":"自定义类名","en":"Custom class name","default":"","version":""},"required":false,"type":"string "},{"name":"style","tag":{"cn":"自定义样式","en":"Custom style","default":"","version":""},"required":false,"type":"CSSProperties "},{"name":"size","tag":{"cn":"不同尺寸","en":"There are three built-in size: small、default、large.","default":"\\\"default\\\"","version":""},"required":false,"type":"\\\"small\\\" | \\\"large\\\" | \\\"default\\\""},{"name":"alt","tag":{"cn":"图像无法显示时的替代文本","en":"The alternative text for the image","default":"","version":""},"required":false,"type":"string "},{"name":"src","tag":{"cn":"图片类头像的资源地址或者图片元素","en":"The address of the image or the image element of the image class avatar","default":"","version":""},"required":false,"type":"ReactNode"},{"name":"icon","tag":{"cn":"设置头像的自定义图标","en":"Set the custom icon of the avatar","default":"","version":""},"required":false,"type":"ReactNode"},{"name":"children","tag":{"cn":"指定头像的形状","en":"Specifies the shape of the avatar","default":"\\\"circle\\\"","version":""},"required":false,"type":"ReactNode"},{"name":"shape","tag":{"cn":"头像文本内容，会根据内容尺寸自适应计算缩放比例","en":"The text content of the avatar, the scaling ratio will be calculated according to the content size","default":"","version":""},"required":false,"type":"\\\"circle\\\" | \\\"square\\\" "},{"name":"srcSet","tag":{"cn":"设置图片类头像响应式资源地址","en":"Set the responsive resource address of the image class avatar","default":"","version":""},"required":false,"type":"string "},{"name":"gap","tag":{"cn":"字符类型距离左右两侧边界单位像素","en":"The distance between the left and right edges of the character type in pixels","default":"","version":""},"required":false,"type":"number "},{"name":"draggable","tag":{"cn":"图片是否允许拖动","en":"Whether the image is allowed to be dragged","default":"true","version":""},"required":false,"type":"boolean "},{"name":"crossOrigin","tag":{"cn":"CORS 属性设置","en":"CORS attribute setting","default":"","version":""},"required":false,"type":"\\\"anonymous\\\" | \\\"use-credentials\\\" "},{"name":"onError","tag":{"cn":"图片加载失败的事件，返回 false 会关闭组件默认的 fallback 行为","en":"The event that the image fails to load, returning false will close the component\\\"s default fallback behavior","default":"","version":""},"required":false,"type":"(() => boolean) "}],"cn":"","en":"","sort":"0"},{"title":"Avatar.Group","properties":[{"name":"className","tag":{"cn":"自定义类名","en":"Custom class name","default":"","version":""},"required":false,"type":"string "},{"name":"style","tag":{"cn":"自定义样式","en":"Custom style","default":"","version":""},"required":false,"type":"CSSProperties "},{"name":"size","tag":{"cn":"不同尺寸","en":"There are three built-in size: small、default、large.","default":"\\\"default\\\"","version":""},"required":false,"type":"\\\"small\\\" | \\\"large\\\" | \\\"default\\\""},{"name":"shape","tag":{"cn":"头像文本内容，会根据内容尺寸自适应计算缩放比例","en":"The text content of the avatar, the scaling ratio will be calculated according to the content size","default":"","version":""},"required":false,"type":"\\\"circle\\\" | \\\"square\\\" "},{"name":"max","tag":{"cn":"设置头像组展示头像的最大数量","en":"Set the maximum number of avatars displayed in the avatar group","default":"","version":""},"required":false,"type":"number "},{"name":"renderMax","tag":{"cn":"自定义渲染隐藏的头像节点，hideAvatarNodes 为隐藏的节点 dom 数组，max 为隐藏的节点数量","en":"Custom rendering of hidden avatar nodes, hideAvatarNodes is an array of hidden node dom, max is the number of hidden nodes","default":"","version":""},"required":false,"type":"((hideAvatarNodes: ReactNode[], max: number) => ReactNode) "},{"name":"children","tag":{"cn":"头像节点","en":"Avatar node","default":"","version":""},"required":true,"type":"ReactNode"},{"name":"popover","tag":{"cn":"用于展示隐藏头像的 Popover 相关配置，配置属性为 Popover 部分属性: position、zIndex、getPopupContainer、adjust","en":"The event that the image fails to load, returning false will close the component\\\"s default fallback behavior","default":"","version":""},"required":false,"type":"Pick<PopoverProps, \\\"position\\\" | \\\"zIndex\\\" | \\\"getPopupContainer\\\" | \\\"adjust\\\"> "}],"cn":"","en":"","sort":"0"}]');
export default api;
