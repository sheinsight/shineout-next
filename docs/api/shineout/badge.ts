const api = JSON.parse('[{"title":"Badge","properties":[{"name":"className","tag":{"cn":"自定义类名","en":"Custom class name","default":"","version":""},"required":false,"type":"string "},{"name":"style","tag":{"cn":"自定义样式","en":"Custom style","default":"","version":""},"required":false,"type":"CSSProperties "},{"name":"children","tag":{"cn":"内容","en":"Content","default":"","version":""},"required":false,"type":"ReactNode"},{"name":"color","tag":{"cn":"自定义徽标颜色","en":"Custom badge color","default":"","version":""},"required":false,"type":"string "},{"name":"count","tag":{"cn":"展示的数字，大于 overflowCount 时显示 ${overflowCount}+，为 0 时隐藏","en":"Number to show, show ${overflowCount}+ when it is greater than overflowCount, hide when it is 0","default":"","version":""},"required":false,"type":"ReactNode"},{"name":"dot","tag":{"cn":"小点模式，开启后不展示数字","en":"Dot mode","default":"false","version":""},"required":false,"type":"boolean "},{"name":"offset","tag":{"cn":"偏移量","en":"Offset of the badge","default":"","version":""},"required":false,"type":"[number, number] "},{"name":"overflowCount","tag":{"cn":"封顶数值","en":"Offset of the badge","default":"","version":""},"required":false,"type":"number "},{"name":"showZero","tag":{"cn":"当数值为 0 时，是否展示 Badge","en":"Whether to show Badge when the number is 0","default":"","version":""},"required":false,"type":"boolean "},{"name":"status","tag":{"cn":"徽标状态","en":"Whether to show Badge when the number is 0","default":"","version":""},"required":false,"type":"\\\"success\\\" | \\\"processing\\\" | \\\"default\\\" | \\\"error\\\" | \\\"warning\\\" "},{"name":"text","tag":{"cn":"状态点文本，仅在配置了 status 属性后生效","en":"Status point text, only valid after configuring the status property","default":"","version":""},"required":false,"type":"ReactNode"},{"name":"size","tag":{"cn":"尺寸，仅在非 dot 模式下生效，支持 small 和 default 两种尺寸","en":"Size, only valid in non-dot mode, supports two sizes: small and default","default":"","version":""},"required":false,"type":"\\\"default\\\" | \\\"small\\\" "}],"cn":"","en":"","sort":"0"}]');
export default api;
