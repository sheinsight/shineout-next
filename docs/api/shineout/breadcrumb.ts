const api = JSON.parse('[{"title":"Breadcrumb","properties":[{"name":"className","tag":{"cn":"自定义类名","en":"Custom class name","default":"","version":""},"required":false,"type":"string "},{"name":"style","tag":{"cn":"自定义样式","en":"Custom style","default":"","version":""},"required":false,"type":"CSSProperties "},{"name":"data","tag":{"cn":"面包屑对象数组,见 data","en":"The array of breadcrumb objects, see data","default":"[]","version":""},"required":false,"type":"any[]"},{"name":"separator","tag":{"cn":"面包屑分隔符,可以是字符串或自定义的元素","en":"A breadcrumb separator which can be strings or custom elements","default":"\\\"/\\\"","version":""},"required":false,"type":"ReactNode"},{"name":"keygen","tag":{"cn":"生成每一项key的辅助方法。为 true 时，以数据项本身作为key，相当于 (d => d)；为函数时，使用此函数返回值；为string时，使用这个string对应的数据值。如 \\\"id\\\"，相当于 (d => d.id)","en":"Key generator.When it is true, the data itself is used as the key equivalent to (d => d);When it is a function, use its return value;When it is a string，ues the value of the string.For example, \\\"id\\\" is the same thing as (d) => d.id.","default":"","version":""},"required":false,"type":"true | ObjectKey<Item> | (data: Item, index?: number ) => string | number"},{"name":"renderItem","tag":{"cn":"自定义渲染","en":"Custom render","default":"","version":""},"required":false,"type":"((value: Item) => ReactNode) "},{"name":"max","tag":{"cn":"最大显示个数","en":"Maximum number of displays","default":"","version":""},"required":false,"type":"number "},{"name":"maxHeight","tag":{"cn":"下拉框最大高度，超出可以内滚","en":"The maximum height of the drop-down box. It can be scrolled if it exceeds the height.","default":"50vh","version":""},"required":false,"type":"string | number "}],"cn":"","en":"","sort":"0"},{"title":"BreadcrumbData","properties":[{"name":"onClick","tag":{"cn":"点击事件","en":"The click event","default":"","version":""},"required":false,"type":"MouseEventHandler<HTMLAnchorElement> "},{"name":"title","tag":{"cn":"显示内容","en":"Displayed content","default":"","version":""},"required":false,"type":"ReactNode"},{"name":"url","tag":{"cn":"链接地址，onClick 属性二选一","en":"Link address","default":"","version":""},"required":false,"type":"string "},{"name":"icon","tag":{"cn":"自定义图标","en":"Custom icon","default":"","version":""},"required":false,"type":"ReactNode"}],"cn":"","en":"","sort":"0"}]');
export default api;
