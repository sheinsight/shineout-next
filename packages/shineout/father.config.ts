import { defineConfig } from 'father';
export default defineConfig({
  extends: '../../.fatherrc.ts',
});
