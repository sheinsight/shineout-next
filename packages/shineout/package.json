{
  "name": "shineout",
  "version": "3.0.0",
  "description": "A components library for React",
  "keywords": [],
  "license": "ISC",
  "author": "",
  "sideEffects": false,
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "typings": "./dist/cjs/index.d.ts",
  "scripts": {
    "build": "father build",
    "build:watch": "father dev",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "@sheinx/base": "workspace:*",
    "@sheinx/hooks": "workspace:*",
    "@sheinx/shineout-style": "workspace:*"
  },
  "peerDependencies": {
    "classnames": ">=2",
    "immer": ">=10",
    "react": ">=16.8"
  },
  "devDependencies": {
    "jszip": "^3.10.1"
  }
}
