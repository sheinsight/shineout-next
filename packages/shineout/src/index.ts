export { default as Alert } from './alert';
export { default as Button } from './button';
export { default as Checkbox } from './checkbox';
export { default as DatePicker } from './date-picker';
export { default as Dropdown } from './dropdown';
export { default as EditableArea } from './editable-area';
export { default as Form } from './form';
export { default as Icon } from './icon';
export { default as Image } from './image';
export { default as Input } from './input';
export { default as Popover } from './popover';
export { default as Radio } from './radio';
export { default as Rule } from './rule';
export { default as Spin } from './spin';
export { default as Switch } from './switch';
export { default as Tag } from './tag';
export { default as Tabs } from './tabs';
export { default as Textarea } from './textarea';
export { default as Tooltip } from './tooltip';
export { setConfig } from '@sheinx/base';

export * as TYPE from './type';
