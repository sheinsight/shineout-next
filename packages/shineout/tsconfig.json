{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "baseUrl": "./",
    "paths": {
      "shineout": ["./src/index"],
      "@sheinx/base": ["../base/src/index"],
      "@sheinx/hooks": ["../hooks/src/index"]
    }
  },
  "include": ["src"]
}
