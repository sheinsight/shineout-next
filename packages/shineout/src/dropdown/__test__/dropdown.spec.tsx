import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import Dropdown from '..';
import { TYPE } from 'shineout';
import mountTest from '../../tests/mountTest';
import { classLengthTest } from '../../tests/structureTest';
import {
  baseTest,
  snapshotTest,
  displayTest,
  attributesTest,
  classTest,
  textContentTest,
  styleTest,
  hasAttributesTest,
  classContentTest,
  styleContentTest,
  delay,
} from '../../tests/utils';
import DropdownBase from '../__example__/001-base';
import DropdownHover from '../__example__/002-hover';
import DropdownPosition from '../__example__/003-position';
import DropdownButton from '../__example__/004-button';
import DropdownChildren from '../__example__/005-children';
import DropdownGroup from '../__example__/006-group';
import DropdownDivider from '../__example__/007-divider';
import DropdownColumns from '../__example__/008-columns';
import DropdownIcon from '../__example__/009-icon';
import DropdownSize from '../__example__/010-size';
import DropdownSplit from '../__example__/011-split';
import DropdownAbsolute from '../__example__/012-absolute';

const SO_PREFIX = 'dropdown';
const dropdownClassName = `.${SO_PREFIX}-wrapper-0-2-1`;
const dropdownButtonClassName = `${SO_PREFIX}-button-0-2-9`;
const dropdownContentClassName = `.${SO_PREFIX}-content-0-2-8`;
const dropdownCaretClassName = `.${SO_PREFIX}-caret-0-2-7`;
const dropdownListClassName = `.${SO_PREFIX}-list-0-2-3`;
const dropdownItemClassName = `${SO_PREFIX}-item-0-2-10`;
const dropdownDisabledClassName = 'button-disabled-0-2-43';
const dropdownOpenClassName = `${SO_PREFIX}-open-0-2-2`;
const dropdownGroupClassName = `.${SO_PREFIX}-optionGroup-0-2-12`;
const dropdownDividerClassName = `.${SO_PREFIX}-optionDivider-0-2-13`;
const dropdownListSmallClassName = `${SO_PREFIX}-listSmall-0-2-4`;
const dropdownListLargeClassName = `${SO_PREFIX}-listLarge-0-2-5`;
const buttonSmallClassName = 'button-small-0-2-27';
const buttonLargeClassName = 'button-large-0-2-28';
const dropdownClassNameOtherClassName = `.${SO_PREFIX}-wrapper-0-2-22`;
const dropdownSplitButtonClassName = `${SO_PREFIX}-splitButton-0-2-36`;
const dropdownListShowClassName = 'animation-list-show-0-2-22';
const animationList = 'animation-list-fade-animation-240-0-2-19';

type DropdownItem = TYPE.Dropdown.Item;
const data: DropdownItem[] = [
  {
    content: 'Submenu',
    children: [
      {
        content: 'Link to Google',
        target: '_blank',
        url: 'https://google.com',
      },
      {
        content: 'Disabled',
        disabled: true,
      },
    ],
  },
  <a key={'link'} href='/'>
    Home
  </a>,
  {
    content: 'Message',
    onClick: () => {
      console.info('Some message.');
    },
  },
];
const menu: DropdownItem[] = [{ content: 'America' }, { content: 'Germany' }];
const dataPosition = 'data-position';
const closeStyle = 'opacity: 0; pointer-events: none; display: none;';

const textSplit = (e: Element) => e.textContent?.split('AmericaGermany')[0];

afterEach(cleanup);
mountTest(<Dropdown data={data} />);
describe('Dropdown[Base]', () => {
  displayTest(Dropdown as React.FC, 'ShineoutDropdown');
  baseTest(
    <Dropdown
      placeholder='Dropdown'
      data={data}
      className='demo'
      style={{ backgroundColor: 'red' }}
    />,
    dropdownClassName,
  );
  snapshotTest(<DropdownBase />);
  snapshotTest(<DropdownHover />, 'about hover');
  snapshotTest(<DropdownPosition />, 'about position');
  snapshotTest(<DropdownButton />, 'about button');
  snapshotTest(<DropdownChildren />, 'about children');
  snapshotTest(<DropdownGroup />, 'about group');
  snapshotTest(<DropdownDivider />, 'about divider');
  snapshotTest(<DropdownColumns />, 'about columns');
  snapshotTest(<DropdownIcon />, 'about icon');
  snapshotTest(<DropdownSize />, 'about size');
  snapshotTest(<DropdownSplit />, 'about split');
  snapshotTest(<DropdownAbsolute />, 'about absolute');
  test('should render default', async () => {
    const childrens = ['Submenu', 'Link to Google', 'Disabled', 'Home', 'Message'];
    const { container } = render(<Dropdown placeholder='Dropdown' data={data} />);
    const dropdown = container.querySelector(dropdownClassName)!;
    expect(dropdown).toBeTruthy();
    attributesTest(dropdown, dataPosition, 'bottom-left');
    classLengthTest(dropdown, 'button', 1);
    const button = dropdown.querySelector('button')!;
    classTest(button, dropdownButtonClassName);
    textContentTest(button.querySelector(dropdownContentClassName)!, 'Dropdown');
    const caret = button.querySelector(dropdownCaretClassName)!;
    classLengthTest(caret, 'svg', 1);
    const list = dropdown.querySelector(dropdownListClassName)!;
    styleTest(list, closeStyle);
    const listDropdown = list.querySelector(dropdownClassName)!;
    attributesTest(listDropdown, dataPosition, 'right-top');
    classTest(listDropdown.querySelector('.' + dropdownButtonClassName)!, dropdownItemClassName);
    classLengthTest(list, 'a', 5);
    list.querySelectorAll('a').forEach((a, index) => {
      classTest(a, dropdownItemClassName);
      textContentTest(a, childrens[index]);
    });
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      classTest(container.querySelectorAll(dropdownClassName)[0]!, dropdownOpenClassName);
      classTest(list, dropdownListShowClassName);
      expect(list.getAttribute('style')).not.toBe(closeStyle);
    });
  });
  test('should render when set disabled is true', () => {
    const clickFn = jest.fn();
    const { container } = render(
      <Dropdown placeholder='Dropdown' data={data} disabled onClick={clickFn} />,
    );
    const button = container.querySelector('button')!;
    classTest(button, dropdownDisabledClassName);
    fireEvent.click(button);
    expect(clickFn.mock.calls.length).toBe(0);
  });
  test('should render when set item disabled is true', () => {
    const clickFn = jest.fn();
    const dataD: DropdownItem[] = [
      {
        content: 'Message',
        onClick: clickFn,
        disabled: true,
      },
      {
        content: 'Message1',
      },
    ];
    const { container } = render(<Dropdown placeholder='Dropdown' data={dataD} />);
    const item = container.querySelectorAll('a')[0];
    hasAttributesTest(item, 'disabled');
    fireEvent.click(item);
    expect(clickFn.mock.calls.length).toBe(0);
  });
});
describe('Dropdown[Hover]', () => {
  test('should render when set hover', async () => {
    const { container } = render(<Dropdown trigger='hover' placeholder='Hover me' data={menu} />);
    const dropdown = container.querySelector(dropdownClassName)!;
    const list = dropdown.querySelector(dropdownListClassName)!;
    styleTest(list, closeStyle);
    fireEvent.mouseEnter(dropdown);
    await waitFor(() => {
      classTest(dropdown, dropdownOpenClassName);
      expect(list.getAttribute('style')).not.toBe(closeStyle);
    });
  });
});
describe('Dropdown[position]', () => {
  const DropdownPositionDemo: React.FC = () => (
    <div style={{ display: 'inline-flex', flexDirection: 'column', gap: 12 }}>
      <div style={{ display: 'flex', justifyContent: 'center', gap: 12, padding: '0 120px' }}>
        <Dropdown placeholder='Bottom left' position='bottom-left' data={menu} />
        <Dropdown placeholder='Bottom' position='bottom' data={menu} />
        <Dropdown placeholder='Bottom right' position='bottom-right' data={menu} />
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Dropdown placeholder='Right Top' position='right-top' data={menu} />
        <Dropdown placeholder='Left Top' position='left-top' data={menu} />
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Dropdown placeholder='Right' position='right' data={menu} />
        <Dropdown placeholder='Left' position='left' data={menu} />
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Dropdown placeholder='Right Bottom' position='right-bottom' data={menu} />
        <Dropdown placeholder='Left Bottom' position='left-bottom' data={menu} />
      </div>
      <div style={{ display: 'flex', justifyContent: 'center', gap: 12 }}>
        <Dropdown placeholder='Top Left' position='top-left' data={menu} />
        <Dropdown placeholder='Top' position='top' data={menu} />
        <Dropdown placeholder='Top right' position='top-right' data={menu} />
      </div>
    </div>
  );
  test('should render when set different position', () => {
    const { container } = render(<DropdownPositionDemo />);
    container.querySelectorAll(dropdownClassName).forEach((dropdown) => {
      attributesTest(
        dropdown,
        dataPosition,
        textSplit(dropdown)?.toLocaleLowerCase().split(' ').join('-') as string,
      );
    });
  });
});
describe('Dropdown[Button]', () => {
  const DropdownButtonDemo: React.FC = () => (
    <div>
      <div className='type'>
        <Dropdown data={menu} type={'primary'} placeholder={'primary'} />
        <Dropdown data={menu} type={'secondary'} placeholder={'secondary'} />
        <Dropdown data={menu} type={'danger'} placeholder={'danger'} />
        <Dropdown data={menu} type={'warning'} placeholder={'warning'} />
        <Dropdown data={menu} type={'success'} placeholder={'success'} />
      </div>
      <div className='other'>
        <Dropdown data={menu} type={'primary'} placeholder={'Outline'} outline />
        <Dropdown data={menu} type={'primary'} placeholder={'Text'} text />
      </div>
    </div>
  );
  test('should render when set button', () => {
    const { container } = render(<DropdownButtonDemo />);
    const type = container.querySelector('.type');
    const dropdowns = type?.querySelectorAll(dropdownClassName);
    dropdowns?.forEach((dropdown) => {
      classContentTest(dropdown.querySelector('button')!, textSplit(dropdown) as string);
    });
  });
  test('should render when set different type', () => {
    const { container } = render(<DropdownButtonDemo />);
    const other = container.querySelector('.other');
    const dropdowns = other?.querySelectorAll(dropdownClassName);
    dropdowns?.forEach((dropdown) => {
      classContentTest(
        dropdown.querySelector('button')!,
        textSplit(dropdown)?.toLocaleLowerCase() as string,
      );
    });
  });
});
describe('Dropdown[Children]', () => {
  test('should render when set group in children', () => {
    const { container } = render(<DropdownGroup />);
    const list = container.querySelector(dropdownListClassName)!;
    const groups = list.querySelectorAll(dropdownGroupClassName)!;
    const items = list.querySelectorAll('a')!;
    expect(groups.length).toBe(items.length / 2);
    for (let i = 0; i < groups.length; i++) {
      textContentTest(groups[i], 'group' + i);
      expect(groups[i]).toBeInTheDocument();
      expect(items[i * 2]).toBeInTheDocument();
      expect(items[i * 2 + 1]).toBeInTheDocument();
    }
  });
  test('should render when set divider in children', () => {
    const { container } = render(<DropdownDivider />);
    const list = container.querySelector(dropdownListClassName)!;
    expect(list.querySelector(dropdownDividerClassName)).toBeInTheDocument();
  });
});
describe('Dropdown[Columns]', () => {
  const width = 500;
  const columns = 5;
  const DropdownColumsDemo = ({ c, w }: { c?: number; w?: number }) => {
    const menu = new Array(30).fill(null).map((_, index) => ({
      id: `${index}`,
      content: `item${index}`,
    }));
    return <Dropdown placeholder='Tiling Menu' width={w} columns={c} data={menu} />;
  };
  test('should render when set columns', () => {
    const { container } = render(<DropdownColumsDemo c={columns} w={width} />);
    const list = container.querySelector(dropdownListClassName)!;
    styleContentTest(list, `width: ${width}px; grid-template-columns: repeat(${columns}, 1fr);`);
  });
  test('should render when set columns without width', () => {
    const { container } = render(<DropdownColumsDemo c={columns} />);
    const list = container.querySelector(dropdownListClassName)!;
    styleContentTest(list, `grid-template-columns: repeat(${columns}, 1fr);`);
  });
});
describe('Dropdown[Icon]', () => {
  test('should render when set icon', () => {
    const { container } = render(<DropdownIcon />);
    const list = container.querySelector(dropdownListClassName)!;
    list.querySelectorAll(dropdownItemClassName).forEach((item) => {
      classLengthTest(item, 'i', 1);
      classTest(item.querySelector('i')!, 'fa-user-o');
    });
  });
});
describe('Dropdown[Size]', () => {
  const buttonClassNames = [buttonSmallClassName, '', buttonLargeClassName];
  const listClassNames = [dropdownListSmallClassName, '', dropdownListLargeClassName];
  test('should render when set size', () => {
    const { container } = render(<DropdownSize />);
    container.querySelectorAll('button').forEach((button, index) => {
      if (index === 1) return;
      classTest(button, buttonClassNames[index]);
    });
    container.querySelectorAll(dropdownListClassName).forEach((list, index) => {
      if (index === 1) return;
      classTest(list, listClassNames[index]);
    });
  });
});
describe('Dropdown[Split]', () => {
  test('should render in group', () => {
    const { container } = render(<DropdownSplit />);
    container.querySelectorAll(dropdownClassNameOtherClassName).forEach((item) => {
      classTest(item.querySelector('button')!, dropdownSplitButtonClassName);
    });
  });
});
describe('Dropdown[absolute]', () => {
  test('should render when set absolute', async () => {
    jest.useRealTimers();
    const { container, rerender } = render(
      <Dropdown type={'primary'} placeholder='Absolute' data={menu} />,
    );
    const dropdown = container.querySelector(dropdownClassName)!;
    expect(dropdown.querySelector(dropdownListClassName)).toBeTruthy();
    rerender(<Dropdown type={'primary'} absolute placeholder='Absolute' data={menu} />);
    expect(dropdown.querySelector(dropdownListClassName)).toBeFalsy();
    expect(document.querySelector(dropdownListClassName)).toBeTruthy();
    styleTest(document.querySelector(dropdownListClassName)!, closeStyle);
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      styleTest(
        document.querySelector(dropdownListClassName)!,
        'display: block; position: absolute; z-index: 1051; min-width: 0; left: 0px; transform: translateX(-100%); top: 2px;',
      );
    });
  });
});
describe('Dropdown[Animation]', () => {
  test('should render when set animation is false', () => {
    const { container } = render(<Dropdown animation={false} data={menu} placeholder='Dropdown' />);
    classTest(container.querySelector(dropdownListClassName)!, animationList, false);
  });
});
describe('Dropdown[OnClick]', () => {
  test('should render when set onClick', async () => {
    const clickFn = jest.fn();
    const { container } = render(<Dropdown onClick={clickFn} data={menu} placeholder='Dropdown' />);
    const dropdown = container.querySelector(dropdownClassName)!;
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      fireEvent.click(dropdown.querySelectorAll('a')[0]!);
      expect(clickFn.mock.calls.length).toBe(1);
    });
  });
  test('should render when have onClick in data', async () => {
    const clickFn = jest.fn();
    const clickChildrenFn = jest.fn();
    const menu: DropdownItem[] = [
      { content: 'America', onClick: clickChildrenFn },
      { content: 'Germany' },
    ];
    const { container } = render(<Dropdown onClick={clickFn} data={menu} placeholder='Dropdown' />);
    const dropdown = container.querySelector(dropdownClassName)!;
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      fireEvent.click(dropdown.querySelectorAll('a')[0]!);
      expect(clickFn.mock.calls.length).toBe(0);
      expect(clickChildrenFn.mock.calls.length).toBe(1);
    });
  });
});
describe('Dropdown[renderItem]', () => {
  const dropData = [
    {
      id: 0,
      content: 'a',
    },
    {
      id: 1,
      content: 'b',
    },
  ];
  test('should render when set renderItem is string', () => {
    const { container } = render(
      <Dropdown renderItem='id' data={dropData} placeholder='Dropdown' />,
    );
    container.querySelectorAll('a').forEach((item, index) => {
      textContentTest(item, String(dropData[index].id));
    });
  });
  test('should render when set renderItem is function', () => {
    const { container } = render(
      <Dropdown renderItem={(d: any) => `id ${d.id}`} data={dropData} placeholder='Dropdown' />,
    );
    container.querySelectorAll('a').forEach((item, index) => {
      textContentTest(item, 'id ' + dropData[index].id);
    });
  });
});
describe('Dropdown[Open]', () => {
  test('should render when set open', () => {
    const { container } = render(<Dropdown open data={menu} placeholder='Dropdown' />);
    const dropdown = container.querySelector(dropdownClassName)!;
    const list = dropdown.querySelector(dropdownListClassName)!;
    classTest(dropdown, dropdownOpenClassName);
    classTest(list, dropdownListShowClassName);
    styleTest(list, 'opacity: 0; pointer-events: none; display: block;');
  });
});
describe('Dropdown[OnCollapse]', () => {
  test('should render when set onCollapes', async () => {
    const collapseFn = jest.fn();
    const { container } = render(
      <Dropdown open data={menu} placeholder='Dropdown' onCollapse={collapseFn} />,
    );
    const dropdown = container.querySelector(dropdownClassName)!;
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      expect(collapseFn.mock.calls.length).toBe(1);
      fireEvent.click(dropdown);
      await delay(200);
      expect(collapseFn.mock.calls.length).toBe(2);
    });
  });
});
describe('Dropdown[Close]', () => {
  test('should close when click button', async () => {
    jest.useFakeTimers();
    const { container } = render(<Dropdown absolute data={menu} placeholder='Dropdown' />);
    const dropdown = container.querySelector(dropdownClassName)!;
    const list = document.querySelector(dropdownListClassName)!;
    styleContentTest(list, 'display: none;');
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      styleContentTest(list, 'display: block;');
    });
    fireEvent.click(dropdown);
    await waitFor(async () => {
      await delay(200);
      styleContentTest(list, 'display: none;');
    });
  });
});
