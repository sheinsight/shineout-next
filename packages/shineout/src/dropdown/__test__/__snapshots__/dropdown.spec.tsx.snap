// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown[Base] should render correctly  1`] = `
<div
  class="dropdown-wrapper-0-2-1"
  data-position="bottom-left"
>
  <button
    class="dropdown-button-0-2-9 button-secondary-0-2-31 button-button-0-2-26 button-text-0-2-38"
  >
    <span
      class="dropdown-content-0-2-8"
    >
      Dropdown
    </span>
    <span
      class="dropdown-caret-0-2-7"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.2929 15.7071C11.6534 16.0676 12.2206 16.0953 12.6129 15.7903L12.7071 15.7071L18.7071 9.70711C19.0976 9.31658 19.0976 8.68342 18.7071 8.29289C18.3466 7.93241 17.7794 7.90468 17.3871 8.2097L17.2929 8.29289L12.7072 12.8779C12.3165 13.2683 11.6835 13.2683 11.2928 12.8779L6.70711 8.29289C6.34662 7.93241 5.77939 7.90468 5.3871 8.2097L5.29289 8.29289C4.93241 8.65338 4.90468 9.22061 5.2097 9.6129L5.29289 9.70711L11.2929 15.7071Z"
        />
      </svg>
    </span>
  </button>
  <div
    class="dropdown-list-0-2-3 animation-list-fade-animation-240-0-2-19 animation-list-fade-0-2-24"
    data-class="dropdown-list-0-2-3 animation-list-fade-animation-240-0-2-19 animation-list-fade-0-2-24"
    style="display: none;"
  >
    <div
      class="dropdown-wrapper-0-2-1"
      data-position="right-top"
    >
      <a
        class="dropdown-button-0-2-9 dropdown-item-0-2-10"
        data-role="item"
      >
        <span
          class="dropdown-content-0-2-8"
        >
          Submenu
        </span>
        <span
          class="dropdown-caret-0-2-7"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.2929 15.7071C11.6534 16.0676 12.2206 16.0953 12.6129 15.7903L12.7071 15.7071L18.7071 9.70711C19.0976 9.31658 19.0976 8.68342 18.7071 8.29289C18.3466 7.93241 17.7794 7.90468 17.3871 8.2097L17.2929 8.29289L12.7072 12.8779C12.3165 13.2683 11.6835 13.2683 11.2928 12.8779L6.70711 8.29289C6.34662 7.93241 5.77939 7.90468 5.3871 8.2097L5.29289 8.29289C4.93241 8.65338 4.90468 9.22061 5.2097 9.6129L5.29289 9.70711L11.2929 15.7071Z"
            />
          </svg>
        </span>
      </a>
      <div
        class="dropdown-list-0-2-3 animation-list-fade-animation-240-0-2-19 animation-list-fade-0-2-24"
        data-class="dropdown-list-0-2-3 animation-list-fade-animation-240-0-2-19 animation-list-fade-0-2-24"
        style="display: none;"
      >
        <a
          class="dropdown-item-0-2-10"
          href="https://google.com"
          target="_blank"
        >
          Link to Google
        </a>
        <a
          class="dropdown-item-0-2-10"
          disabled=""
        >
          Disabled
        </a>
      </div>
    </div>
    <a
      class="dropdown-item-0-2-10"
      href="/"
    >
      Home
    </a>
    <a
      class="dropdown-item-0-2-10"
    >
      Message
    </a>
  </div>
</div>
`;
